src/per/uart.cpp

```diff
@@ -215,106 +481,409 @@ void HAL_UART_AbortReceiveCpltCallback(UART_HandleTypeDef* huart)
 //void HAL_UART_TxCpltCallback(UART_HandleTypeDef *huart);
 //void HAL_UART_RxHalfCpltCallback(UART_HandleTypeDef *huart);
 
+//gotta expand the macros to put them in an arr
+void GPIOA_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOA_CLK_ENABLE();
+}
+void GPIOB_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOB_CLK_ENABLE();
+}
+void GPIOC_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOC_CLK_ENABLE();
+}
+void GPIOD_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOD_CLK_ENABLE();
+}
+void GPIOE_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOE_CLK_ENABLE();
+}
+void GPIOF_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOF_CLK_ENABLE();
+}
+void GPIOG_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOG_CLK_ENABLE();
+}
+void GPIOH_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOH_CLK_ENABLE();
+}
+void GPIOI_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOI_CLK_ENABLE();
+}
+void GPIOJ_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOJ_CLK_ENABLE();
+}
+void GPIOK_CLK_ENABLE()
+{
+    __HAL_RCC_GPIOK_CLK_ENABLE();
+}
+
+typedef void (*VoidFunc)(void);
+
+void GpioClockEnable(GPIO_TypeDef* port)
+{
+    VoidFunc func_p[] = {GPIOA_CLK_ENABLE,
+                         GPIOB_CLK_ENABLE,
+                         GPIOC_CLK_ENABLE,
+                         GPIOD_CLK_ENABLE,
+                         GPIOE_CLK_ENABLE,
+                         GPIOF_CLK_ENABLE,
+                         GPIOG_CLK_ENABLE,
+                         GPIOH_CLK_ENABLE,
+                         GPIOI_CLK_ENABLE,
+                         GPIOJ_CLK_ENABLE,
+                         GPIOK_CLK_ENABLE};
+
+    GPIO_TypeDef* ports[] = {GPIOA,
+                             GPIOB,
+                             GPIOC,
+                             GPIOD,
+                             GPIOE,
+                             GPIOF,
+                             GPIOG,
+                             GPIOH,
+                             GPIOI,
+                             GPIOJ,
+                             GPIOK};
+
+    for(int i = 0; i < 11; i++)
+    {
+        if(port == ports[i])
+        {
+            func_p[i]();
+            return;
+        }
+    }
+}
```

Comment:
This doesn't belong in uart.cpp, it should probably move to the hal_map file so that other peripherals can use it.